File: algorithms/645-set-mismatch/set-mismatch.py
"""
645. set mismatch

https://leetcode.com/problems/set-mismatch/
https://leetcode-cn.com/problems/set-mismatch/

错误的集合
"""

import collections
from typing import List


def find_error_nums_v1(nums: List[int]) -> List[int]:
    """
    哈希表
    """
    length = len(nums)
    counter = collections.Counter(nums)
    ans = [0] * 2
    for i in range(1, length + 1):
        if counter[i] == 2:
            ans[0] = i
        elif counter[i] == 0:
            ans[1] = i
    return ans


def find_error_nums_v2(nums: List[int]) -> List[int]:
    """
    数学方法
    """
    length = len(nums)
    total = sum(set(nums))
    return [sum(nums) - total, (length + 1) * length // 2 - total]

File: algorithms/645-set-mismatch/set-mismatch.go
package problem645

// findErrorNumsV1 哈希表
func findErrorNumsV1(nums []int) []int {
	cnt := map[int]int{}
	for _, num := range nums {
		cnt[num]++
	}
	ans := make([]int, 2)
	for i := 1; i <= len(nums); i++ {
		if c := cnt[i]; c == 2 {
			ans[0] = i
		} else if c == 0 {
			ans[1] = i
		}
	}
	return ans
}

// findErrorNumsV2 数学法
func findErrorNumsV2(nums []int) []int {
	length := len(nums)
	total := sumDistinct(nums)
	ans := make([]int, 2)
	ans[0] = sum(nums) - total
	ans[1] = (1+length)*length/2 - total
	return ans
}

func sum(nums []int) int {
	res := 0
	for _, num := range nums {
		res += num
	}
	return res
}

func sumDistinct(nums []int) int {
	seen := map[int]bool{}
	res := 0
	for _, num := range nums {
		if ok := seen[num]; ok {
			continue
		} else {
			res += num
			seen[num] = true
		}
	}
	return res
}
